docs/Toy Tutorial/toy/parser.py

```diff
@@ -0,0 +1,431 @@
+from pathlib import Path
+from typing import List, NoReturn, TypeVar, cast
+
+from toy.lexer import (OperatorToken, Token, tokenize, NumberToken, EOFToken,
+                       IdentifierToken)
+from toy.ast import (BinaryExprAST, ExprAST, NumberExprAST, FunctionAST,
+                     ModuleAST, LiteralExprAST, VarDeclExprAST, VarType,
+                     VariableExprAST, ReturnExprAST, PrintExprAST, CallExprAST,
+                     PrototypeAST)
+
+
+class ParseError(Exception):
+
+    def __init__(self,
+                 token: Token,
+                 expected: str | type[Token],
+                 context: str = '',
+                 line: str = ''):
+        loc = token.loc
+
+        message = f'Parse error ({loc.line}, {loc.col}): expected '
+        if isinstance(expected, str):
+            message += expected
+        else:
+            message += expected.name()
+
+        if len(context):
+            message += ' ' + context
+
+        message += f" but has Token '{token.text}'\n"
+
+        if len(line):
+            message += line + '\n'
+
+        super().__init__(message)
+
+    pass
+
+
+TokenT = TypeVar('TokenT', bound=Token)
+
+
+class Parser:
```

Comment:
Could we maybe somehow unify this with our parser infrastructure? Or maybe even autogenerate at some point (@AntonLydike).  I guess not for this PR though :)

Reply:
I'm not sure how helpful that would be, this is a sort of "toy" parser, that we don't need to be particularly efficient or good to work on, more for didactic purposes. Might be a bit of a distraction for people to try to understand the parser generation infrastructure instead of xDSL itself